### DISCUSSION TITLE
Various installation issues

### DISCUSSION THREAD

**Discussion by TheLemonPig** [MilesCranmer/PySR#262] [Q&A]
I had been struggling to get PySR working in most environments but was successful in running the Google Colab tutorial. I was hoping to migrate the functionality of the first block of code out of Google Colab into a python or .pytoml file in order to run PySR outside of the Google Colab environment. However I have been struggling understand what the block of code does to make PySR work.

```sh
!echo "Runtime started."

%%shell
set -e

#---------------------------------------------------#
JULIA_VERSION="1.7.2"
JULIA_NUM_THREADS=4
export JULIA_PKG_PRECOMPILE_AUTO=0
#---------------------------------------------------#

if [ -z `which julia` ]; then
  # Install Julia
  JULIA_VER=`cut -d '.' -f -2 <<< "$JULIA_VERSION"`
  echo "Installing Julia $JULIA_VERSION on the current Colab Runtime..."
  BASE_URL="https://julialang-s3.julialang.org/bin/linux/x64"
  URL="$BASE_URL/$JULIA_VER/julia-$JULIA_VERSION-linux-x86_64.tar.gz"
  wget -nv $URL -O /tmp/julia.tar.gz # -nv means "not verbose"
  tar -x -f /tmp/julia.tar.gz -C /usr/local --strip-components 1
  rm /tmp/julia.tar.gz

  echo "Installing PyCall.jl..."
  julia -e 'using Pkg; Pkg.add("PyCall"); Pkg.build("PyCall")'
  julia -e 'println("Success")'

fi
```
Doing this Google Colab specific task was also necessary:
" Use fallback runtime. Run the above cell (!echo "Runtime started.") to start the runtime. Now, open the command pallette (bottom left -> second icon from the bottom), and search for "use fallback runtime", and hit enter. "

**Comment by MilesCranmer** [comment]
This block of code is colab-specific. It installs Julia within Google Colab, because their VM (based on Ubuntu) doesn't include Julia by default. Note that Google Colab runs in the cloud, rather than on your local machine – so even if Julia is installed locally, it still needs to be installed on the colab VM.

> Doing this Google Colab specific task was also necessary:
> " Use fallback runtime. Run the above cell (!echo "Runtime started.") to start the runtime. Now, open the command pallette (bottom left -> second icon from the bottom), and search for "use fallback runtime", and hit enter. "

This actually changed only 2 hours ago! See https://github.com/googlecolab/colabtools/issues/3385. I need to update the colab notebook to remove that part, as you have noted 🙂 

> I had been struggling to get PySR working in most environments but was successful in running the Google Colab tutorial. 

Please share where you are struggling and I can try to help

---

Edit: updated the notebook.

**Reply by TheLemonPig** [reply]
Sure thing. I am a different points on different implementations. For trying to install via pip locally, I am stuck at adding Julia to the PATH. I have added several which I believe should/could be the correct bin, but it does not seem to correct. There's a good chance I added it wrong though.

Does this look right?: /Volumes/Julia-1.8.5/Julia-1.8.app/Contents/Resources/julia/bin
Or this?: /Users/joshuahewson/opt/anaconda3/pkgs/julia-1.8.2-hcb9ed81_0/bin

For context:
-  I am trying to install PySR via pip
- my machine is running on MAC OS using M2 chip
- Julia was installed for the ARM/M-series processor 

---

Trialling on a unix server, I receive a different error (which might be server-related):
When trying to install pysr: 
"ERROR: Unable to load dependent library /gpfs/rt/7.2/opt/julia/1.7.2/bin/../lib/julia/libjulia-internal.so.1
Message:/lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /gpfs/rt/7.2/opt/julia/1.7.2/bin/../lib/julia/libjulia-internal.so.1)
"

---

I have since had success with the docker build though :) but I still want to see if I can get the pip version working locally

**Reply by MilesCranmer** [reply]
I think both of those paths are wrong. The `/Volumes/` path is just where the installation package is unloaded, but IIRC is just a temporary path. The other one looks to be a version that anaconda installed. But there is no version of Julia available on conda that is compatible with M-series chips (until this is done: https://github.com/conda-forge/julia-feedstock/pull/224), so that one shouldn't work either.

I would delete all of those Julia folders and start from scratch. Maybe try juliaup which is my preferred tool: https://github.com/JuliaLang/juliaup:

```bash
curl -fsSL https://install.julialang.org | sh
```

**Reply by TheLemonPig** [reply]
I think that kind of worked! It added a PATH declaration to .zrsch .profile and .bash_profile
When trying to install it still said it couldn't find the path, so I pasted the PATH declaration from .zrsch and then it worked:
```
path=('/Users/joshuahewson/.juliaup/bin' $path)
export PATH
```
However it ran into an error importing PyCall
Does it look like it would be the right bin?

**Reply by MilesCranmer** [reply]
Yeah that looks right.

For PyCall, did it still run? The first time you install the PySR deps it will display an "error" in importing PyCall but then succeed in installing PyCall right after. If the installation errors out though, then it's real. The most common reason for this is your Julia version changed but PyCall was built for the older Julia version – if so, you can delete the `~/.julia/environments` folder and re-run `python -c 'import pysr; pysr.install()'`.

**Reply by TheLemonPig** [reply]
Unfortunately that didn't work. It is maybe worth mentioning that it will update a conda folder when doing this, and will readd it if removed.
<img width="191" alt="Screenshot 2023-02-17 at 2 47 28 PM" src="https://user-images.githubusercontent.com/43144407/219775796-d6179b9a-ff08-477c-ba73-110d05749a28.png">

**Reply by MilesCranmer** [reply]
Hm I need more info. What didn't work? What was the error message? At what step did it fail?

**Reply by TheLemonPig** [reply]
These are the main messages I get:
[ **Info**: Julia executable: /Users/joshuahewson/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/bin/julia
[ **Info**: Trying to import PyCall...
**ERROR**: LoadError: PyCall not properly installed. Please run Pkg.build("PyCall")
**Error**: `import PyCall` failed
 **Info**: PyCall is already installed but importing it failed.
└ Re-building PyCall may fix the issue...
[ **Info**: Run `Pkg.build("PyCall"; verbose=true)`
    Building Conda ─→ `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/e32a90da027ca45d84678b826fffd3110bb3fc90/build.log`
┌ **Info**: CONDA_EXE not found. Reverting to default in ROOTENV
│   CONDA_EXE = "/Users/joshuahewson/.julia/conda/3/x86_64/bin/conda"
└   ROOTENV = "/Users/joshuahewson/.julia/conda/3/x86_64"
    Building PyCall → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/62f417f6ad727987c755549e9cd88c46578da562/build.log`
**ERROR**: LoadError: Error building `PyCall`:

**Reply by MilesCranmer** [reply]
What's after ` ERROR: LoadError: Error building PyCall:`?

**Reply by TheLemonPig** [reply]
**Building** PyCall → `~/.julia/scratchspaces/44cfe95a-1eb2-52ea-b672-e2afdf69b78f/62f417f6ad727987c755549e9cd88c46578da562/build.log`
**ERROR**: LoadError: Couldn't find libpython; check your PYTHON environment variable.

The python executable we tried was /Users/joshuahewson/Developer/venvs/pysrvenv/bin/python (= version 3.10).
Re-building with
    ENV["PYCALL_DEBUG_BUILD"] = "yes"
may provide extra information for why it failed.

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] find_libpython(python::String; _dlopen::typeof(Base.Libc.Libdl.dlopen))
   @ Main ~/.julia/packages/PyCall/twYvK/deps/buildutils.jl:72
 [3] find_libpython(python::String)
   @ Main ~/.julia/packages/PyCall/twYvK/deps/buildutils.jl:40
 [4] top-level scope
   @ ~/.julia/packages/PyCall/twYvK/deps/build.jl:82
 [5] include(fname::String)
   @ Base.MainInclude ./client.jl:476
 [6] top-level scope
   @ none:5
in expression starting at /Users/joshuahewson/.julia/packages/PyCall/twYvK/deps/build.jl:43
**ERROR**: LoadError: Error building `PyCall`: 

Stacktrace:
  [1] pkgerror(msg::String)
    @ Pkg.Types ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Types.jl:67
  [2] (::Pkg.Operations.var"#66#73"{Bool, Pkg.Types.Context, String, Pkg.Types.PackageSpec, String})()
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1060
  [3] withenv(::Pkg.Operations.var"#66#73"{Bool, Pkg.Types.Context, String, Pkg.Types.PackageSpec, String}, ::Pair{String, String}, ::Vararg{Pair{String}})
    @ Base ./env.jl:172
  [4] (::Pkg.Operations.var"#107#112"{String, Bool, Bool, Bool, Pkg.Operations.var"#66#73"{Bool, Pkg.Types.Context, String, Pkg.Types.PackageSpec, String}, Pkg.Types.PackageSpec})()
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1619
  [5] with_temp_env(fn::Pkg.Operations.var"#107#112"{String, Bool, Bool, Bool, Pkg.Operations.var"#66#73"{Bool, Pkg.Types.Context, String, Pkg.Types.PackageSpec, String}, Pkg.Types.PackageSpec}, temp_env::String)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1493
  [6] (::Pkg.Operations.var"#105#110"{Dict{String, Any}, Bool, Bool, Bool, Pkg.Operations.var"#66#73"{Bool, Pkg.Types.Context, String, Pkg.Types.PackageSpec, String}, Pkg.Types.Context, Pkg.Types.PackageSpec, String, Pkg.Types.Project, String})(tmp::String)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1582
  [7] mktempdir(fn::Pkg.Operations.var"#105#110"{Dict{String, Any}, Bool, Bool, Bool, Pkg.Operations.var"#66#73"{Bool, Pkg.Types.Context, String, Pkg.Types.PackageSpec, String}, Pkg.Types.Context, Pkg.Types.PackageSpec, String, Pkg.Types.Project, String}, parent::String; prefix::String)
    @ Base.Filesystem ./file.jl:764
  [8] mktempdir(fn::Function, parent::String) (repeats 2 times)
    @ Base.Filesystem ./file.jl:760
  [9] sandbox(fn::Function, ctx::Pkg.Types.Context, target::Pkg.Types.PackageSpec, target_path::String, sandbox_path::String, sandbox_project_override::Pkg.Types.Project; preferences::Dict{String, Any}, force_latest_compatible_version::Bool, allow_earlier_backwards_compatible_versions::Bool, allow_reresolve::Bool)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1540
 [10] build_versions(ctx::Pkg.Types.Context, uuids::Set{Base.UUID}; verbose::Bool)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:1041
 [11] build(ctx::Pkg.Types.Context, uuids::Set{Base.UUID}, verbose::Bool)
    @ Pkg.Operations ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/Operations.jl:898
 [12] build(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; verbose::Bool, kwargs::Base.Pairs{Symbol, Base.TTY, Tuple{Symbol}, NamedTuple{(:io,), Tuple{Base.TTY}}})
    @ Pkg.API ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/API.jl:1029
 [13] build(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.TTY, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:verbose,), Tuple{Bool}}})
    @ Pkg.API ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/API.jl:156
 [14] #build#99
    @ ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/API.jl:144 [inlined]
 [15] #build#98
    @ ~/.julia/juliaup/julia-1.8.5+0.x64.apple.darwin14/share/julia/stdlib/v1.8/Pkg/src/API.jl:143 [inlined]
 [16] build_pycall()
    @ Main ~/Developer/venvs/pysrvenv/lib/python3.10/site-packages/julia/install.jl:54
 [17] top-level scope
    @ ~/Developer/venvs/pysrvenv/lib/python3.10/site-packages/julia/install.jl:115
in expression starting at /Users/joshuahewson/Developer/venvs/pysrvenv/lib/python3.10/site-packages/julia/install.jl:73
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/joshuahewson/Developer/venvs/pysrvenv/lib/python3.10/site-packages/pysr/julia_helpers.py", line 84, in install
    julia.install(quiet=quiet)
  File "/Users/joshuahewson/Developer/venvs/pysrvenv/lib/python3.10/site-packages/julia/tools.py", line 118, in install
    raise PyCallInstallError("Installing", output)
julia.tools.PyCallInstallError: Installing PyCall failed.

** Important information from Julia may be printed before Python's Traceback **

Some useful information may also be stored in the build log file
`~/.julia/packages/PyCall/*/deps/build.log`.

**Reply by MilesCranmer** [reply]
Very strange...

Can you try to delete your entire `~/.julia` and then re-run the installation?

**Reply by TheLemonPig** [reply]
Same output unfortunately

**Reply by MilesCranmer** [reply]
It looks like the error:
```
 ERROR: LoadError: Couldn't find libpython; check your PYTHON environment variable.
```
relates to it not being able to find `libpython.dylib` or `libpython.so`. Can you check if you have one of those?

**Reply by TheLemonPig** [reply]
I don't seem to have either one of those in my entire computer. Definitely not within the virtual environment I am working :/
